resources/type-system/inference.md

```diff
@@ -299,66 +308,143 @@ The function **futureValueTypeSchema** is defined as follows:
+or if every control path through the block exits the block without
+reaching the end of the block, as computed by the **definite completion**
+analysis specified elsewhere.
+Otherwise _(the function is a non-generator function and at least
+one control path reaches the end of the block)_ let `T` be `null`.
+
+Then process relevant statements of the block, one by one in source order,
+to find a value type `V` for that statement.
+
+  - If the funtion literal is a non-`async` non-generator function,
+    the relevant statements are `return;` or `return e;` statements.
+    * For a `return;` statement, let `V` be `Null`.
+    * For a `return e;` statement, let `V` be the inferred type of `e` with
+      typing context `K`.
+
+  - If the function literal is marked `async`, the relevant statements
+    are `return;` and `return e;` statements.
+    * For a `return;` statement, let `V` be `Null`.
+    * For a `return e;` statement,
+      then let `S` be the inferred type of `e` with typing context
+      `FutureOr<K>`.
+      Let `V` be **flatten**(`S`).
+
+  - If the function literal is marked `sync*`, the relevant statements
+    are `yield e;` or `yield* e;` statement.
+    * For a `yield e;` statement, let `V` be the inferred type of `e` with
+      typing context `K`.
+    * For a `yield* e;` statement, let `S` be the inferred type of `e` with
+      typing context `Iterable<K>`.
+      If `S` implements `Iterable<R>` for some `R`, let `V` be `R`.
+      Otherwise let `V` be `S`.
+      _It is a compile-time error if `S` is not a assignable to
+      `Iterable<Object?>`, so either `S` implements `Iterable`,
+      or it is one of `dynamic` or `Never`._
+
+  - If the function literal is marked `async*`, the relevant statements are
+    `yield e;` or `yield* e;` statements.
+    * For a `yield e;` statement, let `V` be the inferred type of `e` with
+      typing context `K`.
+    * For a `yield* e;` statement, let `S` be the inferred type of `e` with
+      typing context `Stream<K>`.
+      If `S` implements `Stream<R>` for some `R`, let `V` be `R`.
+      Otherwise let `V` be `S`.
+      _It is a compile-time error if `S` is not a assignable to
+      `Stream<Object?>`, so either `S` implements `Stream`,
+      or it is one of `dynamic` or `Never`._
+
+After processing each relevant statement, update `T` to be **UP**(`V`, `T`).
+
+The **actual value type** of the function literal is the value of `T` after
+all relevant statements in the block body have been processed.
+
+Let `T` be the **actual value type** of a function literal as computed above.
 Let `R` be the greatest closure of the typing context `K` as computed above.
 
-With null safety: if `R` is `void`, or the function literal is marked `async`
-and `R` is `FutureOr<void>`, let `S` be `void` (without null-safety: no special
-treatment is applicable to `void`).
+_Then compute the actual value/element type, `S`, to use in the inferred return
```

Comment:
I think this is the first time element type is mentioned in this doc.  If you're going to use it define it, but maybe better to just delete?